### PR TITLE
[CN-864] Add support for custom configs using ConfigMap

### PR DESCRIPTION
**Pull request by SeriyBg**
## User Impact

Added possibility to add custom Hazelcast configuration specified in ConfigMap via `customConfigCmName` field in Hazelcast CR

### CODE REVIEW

**Review by cagric0 — APPROVED** [review]
LGTM!

File: controllers/hazelcast/hazelcast_resources.go [hunk 1]
```diff
@@ -697,9 +697,41 @@ func hazelcastConfig(ctx context.Context, c client.Client, h *hazelcastv1alpha1.
...
+		}
+		hzWrapper := make(map[string]interface{})
+		hzWrapper["hazelcast"] = cfgMap
+		return yaml.Marshal(hzWrapper)
+	}
+
 	return yaml.Marshal(config.HazelcastWrapper{Hazelcast: cfg})
 }
 
+func mergeCustomConfig(cfg *config.Hazelcast, cstCfg map[string]interface{}) error {
```
Comment (semihbkgr) [inline controllers/hazelcast/hazelcast_resources.go:720]: This function basically merges the `config` into the `custom config`. Would be good to rename the function and change the parameters order.

File: test/integration/hazelcast_config_test.go
Comment (semihbkgr) [inline test/integration/hazelcast_config_test.go:1]: We did you create a new test file instead of residing them in `hazelcast_test.go`

**Review by semihbkgr — CHANGES_REQUESTED** [review]
What would happen if user gives incorrect Hazelcast config fields in custom config map. Should we check custom config fields in validation?

File: controllers/hazelcast/hazelcast_resources.go (hunk 1 above)
Comment (SeriyBg) [reply controllers/hazelcast/hazelcast_resources.go:720]: Would you happen to have any suggestions for the name?
Why change the order of the parameters?

File: test/integration/hazelcast_config_test.go
Comment (SeriyBg) [reply test/integration/hazelcast_config_test.go:1]: This test file is focused only on the Hazelcast config. I'm not a big fan of having huge files with everything in it, and think that it's better to separate them for a better readbility

**Comment by SeriyBg** [comment]
> What would happen if user gives incorrect Hazelcast config fields in custom config map. Should we check custom config fields in validation?

The decision on not to validate the custom config. Here you can find the reasoning: https://hazelcast.atlassian.net/wiki/spaces/CN/pages/4513398825/Custom+config+support+for+Hazelcast+CR#Should-we-validate-the-custom-config%3F

File: controllers/hazelcast/hazelcast_resources.go (hunk 1 above)
Comment (semihbkgr) [reply controllers/hazelcast/hazelcast_resources.go:720]: my suggestion: `func mergeConfig(cstCfg map[string]interface{},cfg *config.Hazelcast) error`

**Review by semihbkgr — APPROVED** [review]

**Review by semihbkgr — APPROVED** [review]

**Pull request merged**